### ISSUE TITLE
@Value.Enclosing, @Value.Parameter, and nested implementations

### ISSUE THREAD

**Issue by henryptung** [immutables/immutables#436]
The following source, when compiled using Immutables 2.2.12:

```java
import org.immutables.value.Value;

@Value.Enclosing
public interface TestEnclosingNested {

    @Value.Immutable
    interface Impl extends TestEnclosingNested {

        @Value.Parameter
        int value();

    }
}
```

produces the following generated source snippet:
```java
    /**
     * Construct a new immutable {@code Impl} instance.
     * @param value The value for the {@code value} attribute
     * @return An immutable Impl instance
     */
    public static ImmutableTestEnclosingNested.Impl of(int value) {
      return new Impl(value);
    }
```

As noted by the two different ways to reference `Impl`, the latter reference binds to `TestEnclosingNested.Impl`, instead of `ImmutableTestEnclosingNested.Impl`. This behavior seems to have improved between 2.2.8 and 2.2.12 (there were more similar errors before), but still not perfect.

**Comment by elucash** [comment]
Thank you for reporting this!

**Comment by elucash** [comment]
this problem is fixed in newly released 2.3. Please, let us know if there are other similar problems.
